github-api-call-inventory/README.md

```diff
@@ -0,0 +1,56 @@
+# GitHub API Call Rate Limits
+
+## Inventory of GitHub API calls made during a pipeline run
+
+See [this Jupyter Notebook](./github-api-call-inventory.ipynb).
+
+## Collecting metrics on GitHub API calls
+
+We want to count the number of GitHub API calls, so that we can produce an
+indicator about the average API calls per pipeline run. Such an indicator
+could help us notice if code changes result in an increase in the API calls
+made, and so making it more likely that the call rate limits are reached in
+certain orgs.
+
+Packit and Packit-Service interact with GitHub through ogr, which in turn
+relies on PyGithub.
+
+### Approach 1: Patch PyGithub to inject a metric object
+
+The first thing, that was explored is if PyGithub could be patched in some way
+in order to enable Prometheus metrics collection. This would be possible, by
+patching [Requester] to accept a Prometheus metric object and use it. This
+then could be used to pass down a metrics collector object (for example a
+Counter) from PackitAPI through ogr to PyGithub.
+
+Going through the whole flow:
+
+In packit-service GithubProjects are obtained with
+`ServiceConfig.get_project()`. This method (implemented in
+`packit.config.config.Config`) will need to be updated to take a
+`github_call_counter` object and pass it down to `ogr.get_project()`.
+
+Here `ogr.factory.get_project()` would pass it down to `GithubService` which
+then would pass it down to `GithubProject`, which then would pass it down to
+the PyGithub `Github` object, so that it can be used as [a new property of the
+`Requester`], in order to be able to pass it to the connection, and
+incremented every time [a response is received].
```

Comment:
Sounds good!

Reply:
What do you suggest from the deployment point of view?

* A PyGithub source-git?
* Do the patching inside image-build?

Maybe, we can try to submit this upstream since this might be relevant for others as well... But I would expect if this got refused...

Reply:
I would apply the patches inside the image build, to keep things simple. And yes, they could be submitted upstream, to get community feedback on it.

Reply:
Made a note about this.

Reply:
Thanks!